packages/flutter/lib/src/material/ink_well.dart

```diff
@@ -616,31 +632,53 @@ class _InkResponseState<T extends InkResponse> extends State<T> with AutomaticKe
       return;
     }
     setState(() {
-      _handleFocusUpdate();
+      _updateFocusHighlights();
     });
   }
 
-  void _handleFocusUpdate() {
+  void _updateFocusHighlights() {
     bool showFocus;
     switch (WidgetsBinding.instance.focusManager.highlightMode) {
       case FocusHighlightMode.touch:
         showFocus = false;
         break;
       case FocusHighlightMode.traditional:
-        showFocus = enabled && (Focus.of(context, nullOk: true)?.hasPrimaryFocus ?? false);
+        showFocus = enabled && _hasFocus;
         break;
     }
     updateHighlight(_HighlightType.focus, value: showFocus);
   }
 
+  bool _hasFocus = false;
+  void _handleFocusUpdate(bool hasFocus) {
+    _hasFocus = hasFocus;
+    _updateFocusHighlights();
+    if (widget.onFocusChange != null) {
+      widget.onFocusChange(hasFocus);
+    }
+  }
+
   void _handleTapDown(TapDownDetails details) {
-    final InteractiveInkFeature splash = _createInkFeature(details);
-    _splashes ??= HashSet<InteractiveInkFeature>();
-    _splashes.add(splash);
-    _currentSplash = splash;
+    _startSplash(details: details);
     if (widget.onTapDown != null) {
       widget.onTapDown(details);
     }
+  }
+
+  void _startSplash({TapDownDetails details, BuildContext context}) {
+    assert(details != null || context != null);
+
+    Offset globalPosition;
+    if (context != null) {
+      final RenderBox referenceBox = context.findRenderObject();
+      globalPosition = referenceBox.localToGlobal(referenceBox.paintBounds.center);
```

Comment:
You might assert referenceBox.hasSize here, since we're assuming that the widget has been laid out.